xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,371 @@
+      do {
+        for (Route route : routes) {
+          if (route.getRouteMatch().matches(
+              "/" + args.getMethodDescriptor().getFullMethodName(), asciiHeaders)) {
+            selectedRoute = route;
+            break;
+          }
+        }
+        if (selectedRoute == null) {
+          return Result.forError(
+              Status.UNAVAILABLE.withDescription("Could not find xDS route matching RPC"));
+        }
+        RouteAction action = selectedRoute.getRouteAction();
+        if (action.getCluster() != null) {
+          cluster = action.getCluster();
+        } else if (action.getWeightedCluster() != null) {
+          int totalWeight = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            totalWeight += weightedCluster.getWeight();
+          }
+          int select = random.nextInt(totalWeight);
+          int accumulator = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            accumulator += weightedCluster.getWeight();
+            if (select < accumulator) {
+              cluster = weightedCluster.getName();
+              break;
+            }
+          }
+        }
+      } while (!retainCluster(cluster));
+      final String finalCluster = cluster;
+      class SelectionCompleted implements Runnable {
+        @Override
+        public void run() {
+          releaseCluster(finalCluster);
+        }
+      }
+
+      Map<String, ?> rawServiceConfig =
+          generateServiceConfigWithMethodTimeoutConfig(
+              selectedRoute.getRouteAction().getTimeoutNano());
+      if (logger.isLoggable(XdsLogLevel.INFO)) {
+        logger.log(XdsLogLevel.INFO,
+            "Generated service config (method config):\n{0}", new Gson().toJson(rawServiceConfig));
+      }
+      ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+      Object config = parsedServiceConfig.getConfig();
+      if (config == null) {
+        throw new AssertionError(
+            "Bug: invalid config", parsedServiceConfig.getError().asException());
+      }
+      return
+          Result.newBuilder()
+              .setCallOptions(args.getCallOptions().withOption(CLUSTER_SELECTION_KEY, cluster))
+              .setConfig(config)
+              .setCommittedCallback(new SelectionCompleted())
+              .build();
+    }
+
+    private boolean retainCluster(String cluster) {
+      AtomicInteger refCount = clusterRefs.get(cluster);
+      if (refCount == null) {
+        return false;
+      }
+      int count;
+      do {
+        count = refCount.get();
+        if (count == 0) {
+          return false;
+        }
+      } while (!refCount.compareAndSet(count, count + 1));
+      return true;
+    }
+
+    private void releaseCluster(final String cluster) {
+      int count = clusterRefs.get(cluster).decrementAndGet();
+      if (count == 0) {
+        syncContext.execute(new Runnable() {
```

Comment:
If the cluster is added back by `onConfigChanged()` in between above two lines, it will not `putIfAbsent()` to `clusterRefs`, and will be removed later by this runnable.

Reply:
> If the cluster is added back by onConfigChanged() in between above two lines, it will not putIfAbsent() to clusterRefs

Good point. I'll note that it is a problem even before remove() was moved into the sync context.

It looks like the `for (String cluster : addedClusters) {` is broken because it never calls `retainCluster()` if the cluster is already present; it had had zero references for the clusters in the set so it _must_ acquire a reference for every cluster in `addedClusters`. In this case, `retainCluster()` would return `false` because the reference count couldn't be increased from 0→1.

I'll say that I was bothered that releaseCluster() is passed a string instead of a specific object. That make would make fixing the bug easier. I wished the reference count was for _a specific object_ whose reference was tied 1:1 with the reference count. That object could be `AtomicInteger`, but that seems a bit weird and hard to debug; we may want to extend `AtomicInteger` (or use `AtomicIntegerFieldUpdater`) to have it store the cluster name `String` as well.

If we leave things mostly-alone, then we could fix this by adding the missing `retainCluster()` to the `addedClusters` loop. If `retainCluster()` fails it would do the `put()`. We'd then also change release cluster to use the two-argument remove: `clusterRefs.remove(cluster, atomicInteger)`.

Reply:
Before remove() is moved into the sync context, the race window is small. Yeah, but the problem does exists. But now, I think the fix isn't that complicated as add and remove operations for `clusterRefs` are all done in the sync context now. 

`putIfAbsent()` is no longer necessary (but `clusterRefs` still needs to be a concurrent map), we can just check if `clusterRefs` contains the cluster to be added. If it does, increment its ref count; if not, put `AtomicInteger(1)`. These two steps do not need to be atomic. This fixes the problem of failing to add back clusters before the previous deleted one is removed from `clusterRefs`. 

When trying to remove the cluster with ref count 0 (also in the sync context), we need to check again its ref count is indeed 0. If it is, remove it from `clusterRefs`; if it is not, do nothing. These two steps do not need to be atomic as the the only chance that could increment the ref count from 0 to 1 is receiving a new update that adds the cluster back, which happens in the sync context. This prevents removing the cluster that is added back again after its ref count is 0.

I added a test case (`resolve_raceBetweenClusterReleasedAndResourceUpdateAddBackAgain`) that reproduces this issue and now it is fixed.